autode/opt/coordinates/internals.py

```diff
@@ -234,3 +248,316 @@ def _primitive_type(self):
 class AnyPIC(PIC):
     def _populate_all(self, x: np.ndarray) -> None:
         raise RuntimeError("Cannot populate all on an AnyPIC instance")
+
+
+def build_pic_from_species(
+    mol: "Species",
+) -> AnyPIC:
+    """
+    Build a set of primitives from the species, using the graph as
+    a starting point for the connectivity of the species. Also joins
+    any disjoint parts of the graph, and adds hydrogen bonds to
+    ensure that the primitives are redundant
+
+    Args:
+        mol:
+
+    Returns:
+        (AnyPIC): The set of primitive internals
+    """
+    pic = AnyPIC()
+    core_graph = _get_connected_graph_from_species(mol)
+    _add_bonds_from_species(pic, mol, core_graph)
+    _add_angles_from_species(pic, mol, core_graph)
+    _add_dihedrals_from_species(pic, mol, core_graph)
+    return pic
+
+
+def _get_connected_graph_from_species(mol: "Species") -> "MolecularGraph":
+    """
+    Creates a fully connected graph from the graph of a species, by
+    (1) joining disconnected fragments by their shortest distance,
+    (2) connecting constrained bonds, (3) joining hydrogen bonds,
+    if present.
+
+    Args:
+        mol: A species (must have atoms and graph)
+
+    Returns:
+        (MolecularGraph):
+    """
+    assert mol.graph is not None, "Species must have graph!"
+    core_graph = copy.deepcopy(mol.graph)
+
+    # join hydrogen bonds
+    h_bond_x = ["N", "O", "F", "P", "S", "Cl"]
+    for i, j in itertools.combinations(range(mol.n_atoms), r=2):
+        if (
+            mol.atoms[i].label in h_bond_x
+            and mol.atoms[j].label == "H"
+            or mol.atoms[j].label in h_bond_x
+            and mol.atoms[i].label == "H"
+        ):
+            vdw_sum = mol.atoms[i].vdw_radius + mol.atoms[j].vdw_radius
+            if mol.distance(i, j) < 0.9 * vdw_sum:
+                if not core_graph.has_edge(i, j):
+                    core_graph.add_edge(i, j, pi=False, active=False)
+
+    # join disconnected graph components
+    if not core_graph.is_connected:
+        components = core_graph.get_components()
+        for comp_i, comp_j in itertools.combinations(components, r=2):
+            min_dist = float("inf")
+            min_pair = (-1, -1)
+            for i, j in itertools.product(list(comp_i), list(comp_j)):
+                if mol.distance(i, j) < min_dist:
+                    min_dist = mol.distance(i, j)
+                    min_pair = (i, j)
```

Comment:
@t-young31 Yes, cdist matrix may work better, but I feel like the indexing would be painful (because i would have to store an array of the indices in a meshgrid type format)?

Reply:
good point – I don't think the indexing hassle is worth it. let's leave it as is and maybe come back to it if needed. "premature optimisation is the root of all evil" 😉